sources/api/netdog/src/dns.rs

```diff
@@ -0,0 +1,253 @@
+//! The dns module contains the code necessary to gather DNS settings from config file,
+//! supplementing with DHCP lease if it exists.  It also contains the code necessary to write a
+//! properly formatted `resolv.conf`.
+use crate::lease::LeaseInfo;
+use crate::RESOLV_CONF;
+use rand::prelude::SliceRandom;
+use rand::thread_rng;
+use serde::Deserialize;
+use snafu::ResultExt;
+use std::collections::BTreeSet;
+use std::fmt::Write;
+use std::fs;
+use std::iter::FromIterator;
+use std::net::IpAddr;
+use std::path::Path;
+
+static DNS_CONFIG: &str = "/etc/netdog.toml";
+
+#[derive(Default, Debug, Deserialize, PartialEq)]
+pub(crate) struct DnsSettings {
+    #[serde(rename = "name-servers")]
+    nameservers: Option<BTreeSet<IpAddr>>,
+    #[serde(rename = "search-list")]
+    search: Option<Vec<String>>,
+}
+
+impl DnsSettings {
+    /// Create a DnsSettings from TOML config file, supplementing missing settings with settings
+    /// from DHCP lease if provided.  (In the case of static addressing, a DHCP lease won't exist)
+    pub(crate) fn from_config_or_lease(lease: Option<&LeaseInfo>) -> Result<Self> {
+        let mut settings = Self::from_config()?;
+        if let Some(lease) = lease {
+            settings.merge_lease(lease);
+        }
+        Ok(settings)
+    }
+
+    /// Merge missing DNS settings into `self` using DHCP lease
+    fn merge_lease(&mut self, lease: &LeaseInfo) {
+        if self.nameservers.is_none() {
+            // Randomize name server order, for libc implementations like musl that send
+            // queries to the first N servers.
+            let mut dns_servers: Vec<IpAddr> = lease.dns_servers.clone().into_iter().collect();
+            dns_servers.shuffle(&mut thread_rng());
+
+            self.nameservers = Some(BTreeSet::from_iter(dns_servers.into_iter()));
+        }
```

Comment:
I'd prefer to shuffle the nameservers after the merge, just before writing `resolv.conf` - to guarantee that they're always shuffled. Here we don't shuffle if the nameservers are set via API.